src/docs/arch-block.dox

```diff
@@ -1,9 +1,195 @@
+created on disk and the associated \c allocation_size is written out to the
+metadata file. However the block manager itself isn't created. Only when a btree
+is opened does WiredTiger allocate a block manager. Each btree has it's own
+block manager.
+
+@subsection read Reading files and pages
+
+When an existing btree is opened for the first time, the location of the root
+block is contained in the metadata file \c WiredTiger.wt. The block manager will
+read the block at the location specified and return the page image as a buffer
+to the layer above. This will then be instantiated as a page in memory.
+
+From there subsequent page addresses can be read from the root page and the
+process repeated as required. If a cursor traverses to a page which hasn't been
+read into memory the same process will take place.
+
+@subsection Writing
+
+Two cases exist for writing out data using the block manager, checkpoint and
+eviction. Both are explained in the following two sections. When a page image
+is written out the block manager the \c bm->write API is called. See \c bt_io.c
+for more detail.
+
+@subsubsection Checkpoint
+
+For details on checkpoint at the WiredTiger level see @ref arch-checkpoint.
+
+At the block manager level a checkpoint corresponds with a set of blocks which
+are stored in the file associated with the given URI. Typically each file will
+contain a minimum of two checkpoints. When opening an existing file the latest
+checkpoint is read.
+
+When WiredTiger performs a checkpoint clean pages will not result in additional
+blocks being created and as such existing blocks from older checkpoints may be
+reused, this means that while there may be more than one checkpoint in a file
+the block are not guaranteed to be mutually exclusive. Checkpoints are created
+in depth first order, leaf blocks are created, then the parent blocks. This is
+a requirement as the parent blocks contain the addresses of the leaf blocks.
+
+The block manager doesn't guarantee that calling \c bm->write will result in
+the data being flushed to disk. In the checkpoint scenario WiredTiger will also
+call \c bm->sync which will call the file system dependent flush function.
+
+As a checkpoint progresses it updates its three extent lists, these extent lists
+are written out to disk as part of the checkpoint in blocks. When a checkpoint
+is written the previous checkpoints will be cleaned up if possible, suppose we
+have finished writing out checkpoint C and have checkpoints A and B on disk.
+We only need to retain two checkpoints so we can cleanup checkpoint A.
+Checkpoint A's extent lists get merged into checkpoint B's extent lists. If
+a block, N, is in the alloc list of checkpoint A and the removed discard list
+of checkpoint B the block is considered free. Until a block in an alloc list
+is seen in a discard list the space cannot freed.
+
+@subsubsection Eviction
+
+Eviction also utilizes the block manager, when a page is evicted and contains
```

Comment:
This should be two sentences.

_Eviction also utilizes the block manager. When a page is evicted and contains data that needs to be maintained, logically a block needs to be written._

Reply:
240571f